File: labs/lab4/lab4.2.py
import re
import time

def naive_search(text, pattern):
    occurrences = []
    n = len(text)
    m = len(pattern)

    for i in range(n - m + 1):
        if text[i:i+m] == pattern:
            if (i == 0 or not text[i-1].isalpha()) and (i+m == n or not text[i+m].isalpha()):
                occurrences.append(text[i-m:i+m+m])
    return occurrences



def gorners_scheme(text):
    base = 31
    result = ord(text[0])
    for i in range(len(text) - 1):
        result = result * base + ord(text[i + 1])
    return result

def calculate_hash(text):
    q = 2147483647
    return gorners_scheme(text) % q

def rabin_karp_search(text, pattern):
    occurrences = []
    base = 31
    q = 2147483647
    m = len(pattern)
    n = len(text)
    pattern_hash = calculate_hash(pattern)
    text_hash = calculate_hash(text[0:m])

    for i in range(n - m + 1):
        if text_hash == pattern_hash:
            if (i == 0 or not text[i - 1].isalpha()) and (i + m == n or not text[i + m].isalpha()):
                occurrences.append(text[i:i + m])

        if i < n - m:
            text_hash = ((text_hash - ord(text[i]) * base ** (m - 1)) * base + ord(text[i + m])) % q

    return occurrences


with open("ojegov.txt", "r", encoding="utf-8") as file:
    dictionary = file.read()
    count = 0
    for word in dictionary:
        for sym in word:
            count += 1

    print(count)

pattern = "все"

start_time = time.time()
naive_occurrences = naive_search(dictionary, pattern)
naive_time = time.time() - start_time
print("Наивный алгоритм:")
print(f"Количество вхождений слова '{pattern}' как отдельного слова: {len(naive_occurrences)}")
print("Время выполнения:", naive_time)

start_time = time.time()
rk_occurrences = rabin_karp_search(dictionary, pattern)
rk_time = time.time() - start_time
print("\nАлгоритм Рабина-Карпа:")
print(f"Количество вхождений слова '{pattern}' как отдельного слова: {len(rk_occurrences)}")
print("Время выполнения:", rk_time)

with open("output_lab4.txt", "w", encoding="utf-8") as output_file:
    output_file.write("Наивный алгоритм:\n")
    output_file.write(f"Количество вхождений слова '{pattern}' как отдельного слова: {len(naive_occurrences)}")
    output_file.write("\nНайденные вхождения:\n")
    for index, word in enumerate(naive_occurrences):
        output_file.write(f"Позиция {index}: {word}\n")
    output_file.write(f"Время выполнения: {naive_time} сек\n\n")
    output_file.write("Алгоритм Рабина-Карпа:\n")
    output_file.write(f"Количество вхождений слова '{pattern}' как отдельного слова: {len(rk_occurrences)}")
    output_file.write("\nНайденные вхождения:\n")
    for index, word in enumerate(rk_occurrences):
        output_file.write(f"Позиция {index}: {word}\n")
    output_file.write(f"Время выполнения: {rk_time} сек\n")

File: labs/lab4/lab4.2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>

#define PRIME_BASE 31
#define MODULUS 2147483647

// Function to calculate hash using Horner's scheme
unsigned long gorners_scheme(const char *text) {
    unsigned long result = 0;
    for (int i = 0; text[i] != '\0'; i++) {
        result = (result * PRIME_BASE + text[i]) % MODULUS;
    }
    return result;
}

// Function to perform naive search
int naive_search(const char *text, const char *pattern) {
    int occurrences = 0;
    int n = strlen(text);
    int m = strlen(pattern);

    for (int i = 0; i <= n - m; i++) {
        if (strncmp(text + i, pattern, m) == 0 &&
            (i == 0 || !isalpha(text[i - 1])) &&
            (i + m == n || !isalpha(text[i + m]))) {
            occurrences++;
        }
    }

    return occurrences;
}

// Function to perform Rabin-Karp search
int rabin_karp_search(const char *text, const char *pattern) {
    int occurrences = 0;
    int n = strlen(text);
    int m = strlen(pattern);
    unsigned long pattern_hash = gorners_scheme(pattern);
    unsigned long text_hash = gorners_scheme(text);

    for (int i = 0; i <= n - m; i++) {
        if (text_hash == pattern_hash &&
            (i == 0 || !isalpha(text[i - 1])) &&
            (i + m == n || !isalpha(text[i + m]))) {
            occurrences++;
        }

        if (i < n - m) {
            text_hash = ((text_hash - text[i] * (unsigned long)pow(PRIME_BASE, m - 1)) * PRIME_BASE + text[i + m]) % MODULUS;
        }
    }

    return occurrences;
}

int main() {
    FILE *file = fopen("ojegov.txt", "r");
    if (file == NULL) {
        perror("Error opening file");
        return 1;
    }

    // Read the file into a string (assuming it's not too large)
    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);
    char *dictionary = malloc(file_size + 1);
    fread(dictionary, 1, file_size, file);
    dictionary[file_size] = '\0';
    fclose(file);

    clock_t start, end;
    double cpu_time_used;

    start = clock();
    int naive_occurrences = naive_search(dictionary, "все");
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("Наивный алгоритм:\n");
    printf("Количество вхождений слова 'все' как отдельного слова: %d\n", naive_occurrences);
    printf("Время выполнения: %f\n", cpu_time_used);

    start = clock();
    int rk_occurrences = rabin_karp_search(dictionary, "все");
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("\nАлгоритм Рабина-Карпа:\n");
    printf("Количество вхождений слова 'все' как отдельного слова: %d\n", rk_occurrences);
    printf("Время выполнения: %f\n", cpu_time_used);

    // ... (Code to write results to output_lab4.txt)

    free(dictionary);
    return 0;
}
